### PR TITLE
Using public, private, static, prototype properties in JS

### PR DESCRIPTION
**Pull request by bhaskarmelkani**
## TL;DR;
- Private properties are defined by using 'var' keyword inside the object,its not directly accessible to instances.
- Public properties are defined by using `this.propertyName`, it is accessible to instances.
- Prototype properties are defined by `Classname.prototype.propertyName`, its public and is shared by instances.
- Static properties are defined by `Classname.propertyName`, it can be used without creating instances.
## Username

@[bhaskarmelkani](https://www.twitter.com/bhaskarmelkani)
## Extra

Explaining with example as how to create private, public, static, prototype properties in JS without using ES6.

### CODE REVIEW

**Comment by loverajoel** [comment]
hey use the PR template -> https://github.com/loverajoel/jstips/blob/gh-pages/GIT_TEMPLATE.md

**Comment by bhaskarmelkani** [comment]
Done.

File: _posts/en/2016-xx-xx-public-private-static-prototype-properties.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+---
+layout: post
+
+title: Public, private, static and prototype properties in JS.
+tip-number: xx
+tip-username: bhaskarmelkani
+tip-username-profile: https://www.twitter.com/bhaskarmelkani
+tip-tldr: 
+* **private variables** are declared with the 'var' keyword inside the object, and can only be accessed by private functions and privileged methods.
```
Comment (loverajoel) [inline _posts/en/2016-xx-xx-public-private-static-prototype-properties.md]: please write a short tldr

**Comment by bhaskarmelkani** [comment]
Done, check if LGTY

**Comment by zenopopovici** [comment]
@loverajoel We should try to review all old tips before accepting new ones...

**Comment by loverajoel** [comment]
@zenopopovici sure, under review :dancer:

**Comment by zenopopovici** [comment]
@neighborhood999 @gromgit @kurtextrem @loverajoel @adaniloff @drakmail @sjfkai @pastrop Can you guys look over this? We have also other 3 tips under review. If you've got the time, give us some feedback and :shipit: or not vote.

**Comment by adaniloff** [comment]
@zenopopovici I was making a big post with big explanations, but I think "shorter is better" ahahah :sunglasses:.

So, I think this tips need some big changes : 
- some text error (for ex. : in the description which is almost the same as the one from the source it is from, "example of a person" should be changed).
- I also think it could be easier to read if we alternate one block of text with one block of code 
- then a link to babeljs (https://babeljs.io/repl/) could be used to have the whole code tested
- also I'm not a big fan of this kind of OOP-like programming style, I would rather recommand the "module pattern" (by the way the source is 4 years old, I think it's kinda gross and may be too old to be used as a "tip" since JS has changed a lot since then; but it's just my opinion)
- finally I think this lacks of basic explanations. Clearly, it's a tip for people who are new JS developers, and not explaining things like the differences between privileged / public methods makes me sad. I also think it should be written somewhere that, since it's not real OOP, JS public methods != OOP public methods (JS privileged is more like OOP public then JS public).

Sorry for the long post ! :sweet_potato: 

PS : style too long :(
EDIT : I forgot to mention I'll probably be too busy to work on it atm, but I'll try to do it asap

**Comment by gromgit** [comment]
I'll have to give a :-1:, for several reasons:
1. Only the example code is original, as far as I can tell. The tip's text was plagiarized from the URL referenced at the end, right down to the grammatical errors in the latter. (Immediate "red flags" for me: the `tip-tldr` and the text looked like they were written by two different people, and the text refers to a "person" that's only defined in the original URL's example.) It's one thing to have someone else's words inspire your own, and quite another to make a blatant copy.
2. Hacked-up "class" implementations like this are so 2000's. [ES6 classes](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes) are a proper approximation of the real thing, for those who can't/won't wrap their heads around prototypes.
3. The example is inconsistent with the text. For example, the private var `_sum` is missing a `var` declaration, and `this.add` and `this.show` are actually **privileged** methods; since they access instance variables/methods, neither of them can be defined as public methods.
4. Bad things happen if you accidentally invoke this kind of "class" like a function: without the all-important `new` keyword, [`this`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/this) is set to either the global object or `undefined` (the latter applies if `use strict` is in effect). If the former, `x`, `y`, `add` and `show` suddenly become global entities, overwriting any previous definitions and potentially causing strange logic bugs. ES6 classes don't have this issue.

**Comment by pastrop** [comment]
I agree with the pretty much all the comments by @adaniloff and @gromgit . I don't find this tip particularly useful.   ES6 brings pretty significant changes to the way classes are represented.  I would guess that those who are interested in classical inheritance patterns will be well advised to move to ES6 classes.  
So it is :-1: for me. Also, I would recommend this book to anyone interested in deep diving into ES5 inheritance patterns: http://bdcampbell.net/javascript/book/javascript_the_good_parts.pdf look at the chapter 5

**Comment by zenopopovici** [comment]
Thanks guys, was really my sentiment but I wanted to make sure. Closing. 
@bhaskarmelkani Please take into account the comments. Sorry we have to refuse this.

**Comment by pastrop** [comment]
I got a quick question on submitting tips.  I have the repo forked.  So, when I am submitting am  I supposed to use Commit directly option or Create a new branch and start a pull request?
Thanks!

```
  From: Zeno Popovici <notifications@github.com>
```

 To: loverajoel/jstips jstips@noreply.github.com 
Cc: pastrop pas_trop2003@yahoo.com
 Sent: Friday, March 11, 2016 9:49 AM
 Subject: Re: [jstips] Using public, private, static, prototype properties in JS (#157)

@eighborhood999 @gromgit @kurtextrem @loverajoel @adaniloff @drakmail @sjfkai @pastrop Can you guys look over this? We have also other 3 tips under review. If you've got the time, give us some feedback and  or not vote.—

**Comment by gromgit** [comment]
@pastrop You can't commit directly to the main repo unless you're a collaborator, so you'd have to create pull requests. Creating a separate branch for each pull request helps keeps things in order when you're juggling changes to multiple tips, so I generally do the following:
1. [Update local fork from the main repo](https://help.github.com/articles/syncing-a-fork/)
2. Create and checkout a new branch: `git checkout -b tipXX-update gh-pages`
3. Make and commit edits.
4. Push this branch to my fork.
5. Go to my forked repo page. There should be a **Compare and Pull Request** button next to your new branch. Click it and follow the instructions.

**Comment by pastrop** [comment]
@gromgit, thanks! I think I figured it out, just didn't read the instruction attentively enough for the first time.  I did it all yesterday night, hopefully correctly, will check it again later today

```
  From: gromgit <notifications@github.com>
```

 To: loverajoel/jstips jstips@noreply.github.com 
Cc: pastrop pas_trop2003@yahoo.com
 Sent: Saturday, March 12, 2016 12:48 AM
 Subject: Re: [jstips] Using public, private, static, prototype properties in JS (#157)

@pastrop You can't commit directly to the main repo unless you're a collaborator, so you'd have to create pull requests. Creating a separate branch for each pull request helps keeps things in order when you're juggling changes to multiple tips, so I generally do the following:   
- [Update local fork from the main repo]
- Create and checkout a new branch: git checkout -b tipXX-update gh-pages
- Make and commit edits.
- Push this branch to my fork.
- Go to my forked repo page. There should be a Compare and Pull Request button next to your new branch. Click it and follow the instructions.
